src/test/java/com/microsoft/azure/relay/SendReceiveTest.java

```diff
@@ -203,16 +205,17 @@ public void httpLargePOSTAndLargeResponseTest() throws IOException {
 	}
 	
 	private static CompletableFuture<Void> websocketClient(String msgExpected, String msgToSend) {
-		CompletableFuture<Boolean> receivedReply = new CompletableFuture<Boolean>();
+		AtomicBoolean receivedReply = new AtomicBoolean(false);
 		
 		return client.createConnectionAsync().thenAccept((channel) -> {
 			channel.writeAsync(StringUtil.toBuffer(msgToSend)).join();
 			channel.readAsync().thenAccept((bytesReceived) -> {
 				String msgReceived = new String(bytesReceived.array());
-				receivedReply.complete(true);
+				receivedReply.set(true);
 				assertEquals("Websocket sender did not receive the expected reply.", msgExpected, msgReceived);
+				channel.closeAsync().join();
 			});
-		}).thenRun(() -> assertTrue("Did not receive message from websocket sender.", receivedReply.join()));
```

Comment:
>join [](start = 94, length = 4)

CF<bool>.join was capable of waiting until it got signaled. Is there any chance the AtomicBoolean might not be set yet due to it executing async (and I think nobody waits for it)?

Reply:
this stage only executes after the previous stage where the AtomicBoolean is set, so it should be fine if the previous stage changes to .thenCompose instead. If we use CF.join() then it may just hang here if things actually go wrong.

---
In reply to: [263111405](https://github.com/Azure/azure-relay-java/pull/29#discussion_r263111405) [](ancestors = 263111405)

Reply:
Ok, CF.join could certainly deadlock. Something with a timeout would have been better.

I don't see anyone waiting on the channel.readAsync task tree so I'm still not clear how the thenRun would definitely wait until channel.readAsync + thenAccept runs.

---
In reply to: [263112543](https://github.com/Azure/azure-relay-java/pull/29#discussion_r263112543) [](ancestors = 263112543,263111405)